pkg/controller/config-controller.go

```diff
@@ -0,0 +1,396 @@
+		return nil
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{})
+	if err != nil {
+		return err
+	}
+
+	var config *cdiv1.CDIConfig
+	configs, err := c.configLister.CDIConfigs(namespace).List(selector)
+	if err != nil {
+		return err
+	}
+	for _, conf := range configs {
+		if conf.Name == name {
+			config = conf
+		}
+	}
+
+	if config == nil {
+		runtime.HandleError(errors.Errorf("CDIConfig '%s' in work queue no longer exists", key))
+		return nil
+	}
+
+	var url string
+
+	routes, err := c.routeLister.List(selector)
+	if err != nil {
+		return err
+	}
+	for _, route := range routes {
+		url = getURLFromRoute(route, c.uploadProxyServiceName)
+		if url != "" {
+			break
+		}
+	}
+	if url == "" {
+		ingresses, err := c.ingressLister.List(selector)
+		if err != nil {
+			return err
+		}
+		for _, ing := range ingresses {
+			url = getURLFromIngress(ing, c.uploadProxyServiceName)
+			if url != "" {
+				break
+			}
+		}
+	}
+	//if user overrode the upload proxy url, ignore all routes and ingresses
+	if config.Spec.UploadProxyURLOverride != nil && config.Spec.UploadProxyURLOverride != config.Status.UploadProxyURL {
+		url = *config.Spec.UploadProxyURLOverride
+	}
+	newConfig := config.DeepCopy()
+	// mutate newConfig
+	newConfig.TypeMeta.Kind = "CDIConfig"
+	newConfig.TypeMeta.APIVersion = "cdi.kubevirt.io/v1alpha1"
+	newConfig.Status.UploadProxyURL = &url
+
+	err = updateCDIConfig(c.cdiClientSet, newConfig)
+	if err != nil {
+		runtime.HandleError(errors.Errorf("Error updating CDI Config %s: %s", key, err))
+	}
+
+	return nil
+}
+
+// Run sets up ConfigController state and executes main event loop
+func (c *ConfigController) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer func() {
+		c.queue.ShutDown()
+	}()
+
+	glog.V(3).Infoln("Creating CDI config")
+	if _, err := CreateCDIConfig(c.cdiClientSet, c.configName, ""); err != nil {
+		runtime.HandleError(err)
+		return errors.Wrap(err, "Error creating CDI config")
+	}
+
+	glog.V(3).Infoln("Starting config controller Run loop")
+	if threadiness < 1 {
+		return errors.Errorf("expected >0 threads, got %d", threadiness)
+	}
+
+	if ok := cache.WaitForCacheSync(stopCh, c.ingressesSynced, c.routesSynced); !ok {
+		return errors.New("failed to wait for caches to sync")
+	}
+
+	glog.V(3).Infoln("ConfigController cache has synced")
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	glog.Info("Started workers")
+	<-stopCh
+	glog.Info("Shutting down workers")
+	return nil
+}
+
+func (c *ConfigController) createCDIConfig(name string) error {
+
+	var url string
+	routes, err := c.routeLister.List(labels.NewSelector())
+	if err != nil {
+		return err
+	}
+	for _, route := range routes {
+		url = getURLFromRoute(route, c.uploadProxyServiceName)
+		if url != "" {
+			break
+		}
+	}
+	if url == "" {
+		ingresses, err := c.client.ExtensionsV1beta1().Ingresses(metav1.NamespaceAll).List(metav1.ListOptions{
```

Comment:
Can't use the ingress lister here like in `syncHandler`?

Reply:
done